### PR TITLE
Feat: add triggers endpoints

### LINKED ISSUE #114: Support for triggers
# Feature request

## Is your feature request related to a problem? Please describe.

We should implement `trigger` support. This is something we will need to build into our Dashboard

## Describe the solution you'd like

A clean wrapper around `select * from information_schema.triggers`

### PR DESCRIPTION
**Pull request by kiwicopple**
## What kind of change does this PR introduce?

Fix #114

### CODE REVIEW

**Comment by w3b6x9** [comment]
Adding the following list to keep track of development.

PR Status:

- [x] GET /triggers
- [ ] GET /triggers?name=test_trigger&table_name=users
- [ ] POST /triggers
- [ ] DELETE /triggers?name=test_trigger&table_name=users
- [ ] PATCH /triggers
       Trigger properties that can be changed
    - [ ] Name
    - [ ] Extension dependence
    - [ ] Enable/Disable

File: src/lib/sql/triggers.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  is_t.action_orientation AS orientation,
+  is_t.action_timing AS timing,
+  CASE
+    WHEN pg_t.tgenabled = 'D' THEN FALSE
+    ELSE TRUE
+  END AS is_enabled
+FROM
+  pg_trigger AS pg_t
+JOIN
+  pg_class AS pg_c
```
Comment (soedirgo) [inline src/lib/sql/triggers.sql:19]: 💯 nice one on using the system catalog btw!

**Comment by w3b6x9** [comment]
The following are examples to test `/triggers` endpoints from 987f84fb6452391b4ca460dd709ffeb56cb720f0:

Setup:

```
create table public.users_audit (
    id BIGINT generated by DEFAULT as identity,
    created_at timestamptz DEFAULT now(),
    user_id bigint,
    previous_value jsonb
);
```

```
create function public.audit_action()
returns trigger as $$
begin
    insert into public.users_audit (user_id, previous_value)
    values (old.id, row_to_json(old));

    return new;
end;
$$ language plpgsql;
```

1. `POST` `/triggers`

Body: 

```
{
    "name": "check_update",
    "table": "public.users",
    "timing": "before",
    "event": "update",
    "function_name": "public.audit_action",
    "orientation": "row"
}
```

2. `GET` `/triggers`

3. `GET` `/triggers/:id` (Grab the id from previous `GET` response)

4. `PATCH` `/triggers/:id`

Body: 

```
{
    "name": "check_update_update",
    "is_enabled": "false"
}
```

5. `DELETE` `/triggers/:id`

File: src/lib/PostgresMetaTriggers.ts [hunk 2]
```diff
@@ -10,59 +10,272 @@ export default class PostgresMetaTriggers {
...
+      const splitIdx = schema.length
+      if (schemaStartIdx === 0 && name[splitIdx] === '.') {
+        names.push(name.slice(0, splitIdx))
+        names.push(name.slice(splitIdx + 1))
+        break
+      }
+    }
+
+    return names[0] ? names : [name]
+  }
```
Comment (w3b6x9) [inline src/lib/PostgresMetaTriggers.ts:278]: @soedirgo this is something we discussed a few days ago about escaping qualified table names. Even though this has to make an additional query for existing schemas, thought it covers the edge case when schema name itself has period(s). It doesn't cover every edge case of qualified table names but I think it's more than good enough for now. Let me know what you think!

**Comment by w3b6x9** [comment]
@kiwicopple @soedirgo I have a bit more work to follow up https://github.com/supabase/postgres-meta/commit/987f84fb6452391b4ca460dd709ffeb56cb720f0 but wanted to push up my progress so everyone can play around with the endpoints and provide feedback.

File: src/lib/PostgresMetaTriggers.ts [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+      if (startIdx === 0 && fullName[splitIdx] === '.') {
+        names.push(fullName.slice(0, splitIdx))
+        names.push(fullName.slice(splitIdx + 1))
+        break
+      }
+    }
+
+    return names[0] ? names : [fullName]
+  }
+}
```
Comment (w3b6x9) [inline src/lib/PostgresMetaTriggers.ts:286]: @soedirgo this is something we discussed a few days ago about escaping qualified table names. Even though this has to make an additional query for existing schemas, thought it covers the edge case when schema name itself has period(s). It doesn't cover every edge case of qualified table names but I think it's more than good enough for now. Let me know what you think!

File: test/integration/index.spec.js [hunk 4]
```diff
@@ -779,3 +779,184 @@ describe('/publications FOR ALL TABLES', () => {
     assert.equal(stillExists, false)
   })
 })
+
+describe('/triggers', () => {
+  const trigger = {
+      name: 'test_trigger',
+      table: 'public.users_audit',
```
Comment (kiwicopple) [inline test/integration/index.spec.js:786]: I see you have a fully qualified function in the triggers logic, but I think on all the other routes we use `table` and `schema`. I think that's a bit cleaner

Comment (w3b6x9) [reply test/integration/index.spec.js:786]: @kiwicopple sounds good, I'll go and make the changes.

**Review by soedirgo — COMMENTED** [review]
@w3b6x9 left some comments. Bit of a chore to slog through, sorry about that... Let me know if you need some help, esp. for consuming the system catalog, it can be kinda finnicky sometimes.

File: src/lib/PostgresMetaTriggers.ts [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+  constructor(query: (sql: string) => Promise<PostgresMetaResult<any>>) {
+    this.query = query
+  }
+
+  async list(): Promise<PostgresMetaResult<PostgresTrigger[]>> {
+    return await this.query(enrichedTriggersSql)
+  }
+
+  async retrieve({ id }: { id: number }): Promise<PostgresMetaResult<PostgresTrigger[]>>
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:17]: So the reason I made this `PostgresTrigger[]` is because of something you caught and mentioned in one of the comments:

> we can have rows with duplicate ids

Wasn't sure at the time how I wanted to handle it but after speaking with @MildTomato I've made the change to return a single record where `events` is an array (e.g. ['INSERT', 'UPDATE]).

File: src/lib/PostgresMetaTriggers.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+    if (name && schema && table) {
+      const sql = `${enrichedTriggersSql} WHERE triggers.name = ${literal(
+        name
+      )} AND triggers.schema = ${literal(schema)} AND triggers.table = ${literal(table)};`
+
+      return await this.query(sql)
+    }
+
+    return { data: null, error: { message: 'Missing id or name, schema, and table' } }
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:52]: @soedirgo yep, sounds good!

File: src/lib/PostgresMetaTriggers.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+    return { data: triggerRecords!, error: null }
+  }
+
+  private async fetchTriggersById(id: number): Promise<PostgresMetaResult<PostgresTrigger[]>> {
+    const { data: triggers, error } = await this.retrieve({ id })
+    const triggerRecord = Array.isArray(triggers) && triggers[0]
+
+    if (error || !triggerRecord) {
+      return { data: null, error: error || { message: `Trigger with id ${id} does not exist` } }
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:266]: @soedirgo gotcha!

File: src/lib/PostgresMetaTriggers.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+      if (error) {
+        return { data: null, error }
+      }
+    }
+
+    return { data: triggerRecords!, error: null }
+  }
+
+  private async fetchTriggersById(id: number): Promise<PostgresMetaResult<PostgresTrigger[]>> {
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:261]: @w3b6x9 I was also using it `update` as well but I've done a bit of refactoring due to some of your other comments so this will be removed.

File: src/lib/sql/triggers.sql [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  pg_t.oid AS id,
+  is_t.trigger_schema AS schema,
+  is_t.trigger_name AS name,
+  is_t.event_manipulation AS event,
+  is_t.event_object_catalog AS catalog,
+  is_t.event_object_table AS table,
+  is_t.action_condition AS condition,
+  is_t.action_statement AS function_statement,
+  is_t.action_orientation AS orientation,
+  is_t.action_timing AS timing,
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/sql/triggers.sql:11]: @soedirgo

> The primary reason originally was performance, but without taking that into account (I don't think this SQL will be a bottleneck or anything) it's still better to only use the Postgres system catalogs directly, since infoschema is a TSQL standard (?) and often contain information that doesn't reflect Postgres internals very well

Appreciate the background and explanation!

I think we should make an exception, at least for the initial version and can go back and refactor, b/c this will get a bit complicated. `pg_trigger` only contains `tgtype` (bit mask e.g. 20) and we'll have to come up with a query like [this](https://stackoverflow.com/a/55000171).

> event_manipulation events are stored per-row (i.e. one row for each of INSERT, UPDATE, etc.) whereas pg_trigger stores them all in one row (in pg_t.tgtype). So using this, we can have rows with duplicate ids. We should instead represent this as an array.

Really good catch! I mentioned this elsewhere but I decided to just return one record and group the events into an array like you mentioned.

> I find it easier to consume if it's a function name

Agreed, I've updated PR to reflect this.

File: src/lib/PostgresMetaTriggers.ts [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+      timing,
+      event,
+      orientation,
+      condition,
+      transition_relation_reference,
+      function_arguments,
+      constraint,
+      referenced_schema = 'public',
+      referenced_table,
+      default_timing,
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:96]: @soedirgo yea I started with `create`  and tried to get all options in there even though we won't be using some of it in the short term. My initial thought was that others in the community using pg-meta can at least have access to the full range of trigger options. But I think you're right and once demand is there we can always add it in. I'm gonna take out a couple of things like `constraint`, `referenced_schema`, etc. out of this PR. Also, it's not like we're returning those additional options in the trigger sql query anyways.

File: src/lib/PostgresMetaTriggers.ts [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+   * @param {Object} obj - An object.
+   * @param {string} obj.name - Trigger name.
+   * @param {string} obj.schema - Schema name.
+   * @param {string} obj.table - Unqualified table, view, or foreign table name that trigger was created for.
+   * @param {string} obj.function_name - Name of the function to execute. For example: 'public.audit_action'
+   * @param {('before'|'after'|'instead of')} obj.timing - Determines when function is called
+   * during event occurrence.
+   * @param {string} obj.event - Event that will fire the trigger. One of the following 'insert' | 'update' | 'update
+   * of column_name1,column_name2' | 'delete' | 'truncate' or multiple events using 'or' (For example: 'insert or update')
+   * @param {('row'|'statement')} obj.orientation - Trigger function for every row affected by event or
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:67]: @soedirgo 

> For timing and orientation: probably should make the casing consistent with retrieve.

Sounds good!

> I was thinking of making this an array

I agree and will include this change in the PR.

File: src/lib/PostgresMetaTriggers.ts [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+      schema,
+    })
+  }
+
+  async update(
+    id: number,
+    {
+      name: newName,
+      extension,
+      is_extension_dependent,
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:183]: @soedirgo yep, taken out!

File: src/lib/PostgresMetaTriggers.ts [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+      is_enabled,
+      enable_mode,
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:185]: @soedirgo Absolutely agreed! I decided to go with the latter.

File: src/lib/PostgresMetaTriggers.ts [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+          } DEPENDS ON EXTENSION ${ident(extension)};`
+        : ''
+
+    if (is_enabled === 'false') {
+      enableDisableSql = `ALTER TABLE ${qualifiedTableName} DISABLE TRIGGER ${ident(currentName)};`
+    }
+
+    if (is_enabled === 'true') {
+      const enableMode =
+        (enable_mode && ['replica', 'always'].includes(enable_mode) && ident(enable_mode)) || ''
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:219]: @soedirgo yep, included!

File: src/lib/PostgresMetaTriggers.ts [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    const triggerRecord = Array.isArray(triggers) && triggers[0]
+
+    if (error || !triggerRecord) {
+      return { data: null, error: error || { message: `Trigger with id ${id} does not exist` } }
+    }
+
+    return { data: triggers!, error: null }
+  }
+
+  private splitQualifiedName(schemas: string[], fullName: string): string[] {
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:272]: @soedirgo yea, I wasn't sure at the time about function names but I've updated triggers sql with returns of `function_schema` and `function_name` so I'll be dropping `splitQualifiedName` altogether.

File: src/lib/types.ts [hunk 16]
```diff
@@ -113,6 +113,20 @@ export interface PostgresPublication {
   tables: { id: number; name: string; schema: string }[] | null
 }
 
+export interface PostgresTrigger {
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/types.ts:116]: @soedirgo let me wrap up the changes and then I can go deal with these conflicts

File: src/lib/PostgresMetaTriggers.ts [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+      default_timing?: string
+    },
+    { pgSchemas }: { pgSchemas: PostgresMetaSchemas }
+  ): Promise<PostgresMetaResult<PostgresTrigger[]>> {
+    let missingRequiredParam
+    if (!name) missingRequiredParam = 'name'
+    if (!table) missingRequiredParam = 'table'
+    if (!function_name) missingRequiredParam = 'function_name'
+    if (!timing) missingRequiredParam = 'timing'
+    if (!event) missingRequiredParam = 'event'
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:120]: @soedirgo yea, so the reason I did this was for better error messaging. If I didn't include these checks and I missed a param, then the error message is `"SQL identifier cannot be null or undefined"`. When the checks are included then I get `"Missing required param activation"` for example.

File: src/lib/PostgresMetaTriggers.ts [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+      orientation?: string
+      condition?: string
+      transition_relation_reference?: string
+      function_arguments?: string
+      constraint?: string
+      referenced_schema?: string
+      referenced_table?: string
+      default_timing?: string
+    },
+    { pgSchemas }: { pgSchemas: PostgresMetaSchemas }
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:113]: @soedirgo not necessary anymore since I removed `splitQualifiedName` .

File: src/lib/sql/triggers.sql (hunk 9 above)
[the remark this answers is not in the document]
Comment (soedirgo) [reply src/lib/sql/triggers.sql:11]: > I think we should make an exception, at least for the initial version and can go back and refactor

Gotcha, sounds good!

File: src/lib/PostgresMetaTriggers.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:120]: I'm going to remove the missing params validation for now and we can figure out the best way to do it later on.

File: src/lib/PostgresMetaTriggers.ts [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    transition_relation_reference?: string
+    function_args?: string[]
+  }): Promise<PostgresMetaResult<PostgresTrigger>> {
+    const qualifiedTableName = `${ident(schema)}.${ident(table)}`
+    const qualifiedFunctionName = `${ident(function_schema)}.${ident(function_name)}`
+    
+    const triggerOrientation = orientation ? `FOR EACH ${format.string(orientation)}` : ''
+    const triggerCondition = condition ? `WHEN ${format.string(condition)}` : ''
+    const triggerEvents = Array.isArray(events) ? `${format.string(events.join(' OR '))}` : ''
+    const functionArgs = Array.isArray(function_args) ? literal(`${function_args.join(',')}`) : ''
```
Comment (soedirgo) [inline src/lib/PostgresMetaTriggers.ts:136]: I think I misunderstood how the args are used here - should you be escaping the individual arguments *then* join them, or the other way around?

File: src/lib/PostgresMetaTriggers.ts [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    name: string
+    table: string
+    function_name: string
+    activation: string
+    events: string[]
+    function_schema?: string
+    schema?: string
+    orientation?: string
+    condition?: string
+    transition_relation_reference?: string
```
Comment (soedirgo) [inline src/lib/PostgresMetaTriggers.ts:127]: Should we be keeping `transition_relation_reference`? Not sure how often it's used, but if we were to keep it we should add them to retrieve (from `pg_t.tgoldtable` and `pg_t.tgnewtable`).

File: src/lib/PostgresMetaTriggers.ts [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+    function_args?: string[]
+  }): Promise<PostgresMetaResult<PostgresTrigger>> {
+    const qualifiedTableName = `${ident(schema)}.${ident(table)}`
+    const qualifiedFunctionName = `${ident(function_schema)}.${ident(function_name)}`
+    
+    const triggerOrientation = orientation ? `FOR EACH ${format.string(orientation)}` : ''
+    const triggerCondition = condition ? `WHEN ${format.string(condition)}` : ''
```
Comment (soedirgo) [inline src/lib/PostgresMetaTriggers.ts:134]: Nitpick: I think it's better to omit the parens in the argument, i.e. treat it like `function_args`:
```suggestion
    const triggerCondition = condition ? `WHEN ( ${format.string(condition)} )` : ''
```

File: src/lib/types.ts [hunk 22]
```diff
@@ -150,6 +150,34 @@ export const postgresPublicationSchema = Type.Object({
...
+  enabled_mode: Type.Union([
+    Type.Literal('ORIGIN'),
+    Type.Literal('REPLICA'),
+    Type.Literal('ALWAYS'),
+    Type.Literal('DISABLED'),
+  ]),
+  name: Type.String(),
+  table: Type.String(),
+  schema: Type.String(),
+  condition: Type.String(),
```
Comment (soedirgo) [inline src/lib/types.ts:164]: Sorry I wasn't very clear, in the prev. review I meant that `condition?:` should be `condition:`, but it should still be nullable to indicate that there's no condition.

File: src/lib/sql/triggers.sql [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+  END AS enabled_mode,
+  STRING_TO_ARRAY(
+    TRIM(
+      TRAILING '\000' FROM ENCODE(pg_t.tgargs, 'escape')
+    ), ','
+  ) AS function_args,
+  is_t.trigger_name AS name,
+  is_t.event_object_table AS table,
+  is_t.event_object_schema AS schema,
+  COALESCE(is_t.action_condition, '') AS condition,
```
Comment (soedirgo) [inline src/lib/sql/triggers.sql:17]: As per the comment in `types.ts`, this doesn't need to be coalesced

**Review by soedirgo — COMMENTED** [review]
Looks good, just a few more small changes and I think this is ready to be merged :)

File: src/lib/types.ts (hunk 22 above)
Comment (w3b6x9) [reply src/lib/types.ts:164]: @soedirgo no i understood what you meant i just thought that it was better that condition always return a string instead. let me go and update

File: src/lib/PostgresMetaTriggers.ts (hunk 20 above)
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:127]: @soedirgo i'll remove this as well

File: src/lib/PostgresMetaTriggers.ts (hunk 19 above)
Comment (w3b6x9) [reply src/lib/PostgresMetaTriggers.ts:136]: @soedirgo yep, you're right it's the former.

**Review by soedirgo — APPROVED** [review]

**Comment by w3b6x9** [comment]
@soedirgo thanks so much for the review!
